yasa/evaluation.py

```diff
@@ -115,123 +109,403 @@ class EpochByEpochEvaluation:
-                n_trimmed = n_ref - n_test
-                warn_msg = f"`hypno_ref` longer than `hypno_test`, trimmed to {n_test} epochs"
-            else:
-                hypno_test = Hypnogram(hypno_test.hypno[:n_ref], n_stages=hypno_test.n_stages)
-                n_trimmed = n_test - n_ref
-                warn_msg = f"`hypno_test` longer than `hypno_ref`, {n_trimmed} epochs trimmed"
-            ## Q: Should be downplayed as INFO?
-            logger.warning(warn_msg)
-        self.hypno_ref = hypno_ref
-        self.hypno_test = hypno_test
-
-    def get_agreement(self):
+
+    def __init__(self, refr_hyps, test_hyps):
+        from yasa.hypno import Hypnogram  # Avoiding circular import
+
+        assert hasattr(refr_hyps, "__iter__"), "`refr_hyps` must be a an iterable"
+        assert hasattr(test_hyps, "__iter__"), "`test_hyps` must be a an iterable"
+        assert type(refr_hyps) == type(test_hyps), "`refr_hyps` and `test_hyps` must be same type"
+        assert len(refr_hyps) == len(
+            test_hyps
+        ), "`refr_hyps` and `test_hyps` must have the same number of hypnograms"
+
+        if isinstance(refr_hyps, dict):
+            # If user provides dictionaries, split into sleep IDs and hypnograms
+            assert (
+                refr_hyps.keys() == test_hyps.keys()
+            ), "hypnograms in `refr_hyps` and `test_hyps` must have identical sleep IDs"
+            sleep_ids, refr_hyps = zip(*refr_hyps.items())
+            test_hyps = tuple(test_hyps.values())
+        else:
+            # Create hypnogram_ids
+            sleep_ids = tuple(range(1, 1 + len(refr_hyps)))
+
+        assert all(
+            isinstance(hyp, Hypnogram) for hyp in refr_hyps + test_hyps
+        ), "`refr_hyps` and `test_hyps` must only include YASA hypnograms"
+        assert all(
+            h.scorer is not None for h in refr_hyps + test_hyps
+        ), "all hypnograms must have a scorer name"
+        for h1, h2 in zip((refr_hyps + test_hyps)[:-1], (refr_hyps + test_hyps)[1:]):
+            assert h1.labels == h2.labels, "all hypnograms must have the same labels"
+            assert h1.mapping == h2.mapping, "all hypnograms must have the same mapping"
+            assert h1.n_stages == h2.n_stages, "all hypnograms must have the same n_stages"
+        assert all(
+            h1.scorer == h2.scorer for h1, h2 in zip(refr_hyps[:-1], refr_hyps[1:])
+        ), "all `refr_hyps` must have the same scorer"
+        assert all(
+            h1.scorer == h2.scorer for h1, h2 in zip(test_hyps[:-1], test_hyps[1:])
+        ), "all `test_hyps` must have the same scorer"
+        assert all(
+            h1.scorer != h2.scorer for h1, h2 in zip(refr_hyps, test_hyps)
+        ), "each `refr_hyps` and `test_hyps` pair must have unique scorers"
+        assert all(
+            h1.n_epochs == h2.n_epochs for h1, h2 in zip(refr_hyps, test_hyps)
+        ), "each `refr_hyps` and `test_hyps` pair must have the same n_epochs"
```

Comment:
Love all these assertion checks!

Reply:
Ha ya at one point it hit me that these new classes are just assertions and pandas chaining. That's pretty much all the code :/